### PR TITLE
Improve symbolic_script gradients for addmm

### PR DESCRIPTION
**Pull request by ngimel**
This PR modifies symbolic script gradients for addmm to return gradient with the same striding as corresponding input (e.g. if an input is column-major, the gradient would also be column-major). 
Notes:

- This approach creates if-blocks in the backward that are not constant-propagated away. This could be avoided by specializing backward, but, per @mruberry, this exposes another bug with gradient accumulator
- The created if-blocks would prevent TreeReduce optimization fusing multiple `mm` operations, which would hurt LSTM performance, so for now column-major optimization is enabled for `addmm` only, and not for `mm` 

Test plan: added a test checking that returned gradients have expected strides.

### CODE REVIEW

File: torch/csrc/jit/symbolic_script.cpp [hunk 1]
```diff
@@ -1371,6 +1357,29 @@ const std::vector<std::string> functions = {
                 return grad_output, None
             return torch.remainder(self, other), backward
 
+        def AD_column_major(mat):
+            return mat.stride(0)==1 and mat.stride(1) == mat.size(0)
+
+        def AD_mm_backward_self(grad, mat2, mat1):
```
Comment (mruberry) [inline torch/csrc/jit/symbolic_script.cpp]: Should this be AD_mm_backward_mat1(grad, mat1, mat2)? Currently there's a name mismatch with the addmm implementation and inconsistent argument ordering. For example, AD_mm_backward_self actually computes the gradient of mat1, not self.

File: torch/csrc/jit/symbolic_script.cpp [hunk 2]
```diff
@@ -1371,6 +1357,29 @@ const std::vector<std::string> functions = {
...
+            if AD_column_major(mat1):
```
Comment (mruberry) [inline torch/csrc/jit/symbolic_script.cpp]: Should this error if mat1 is sparse like the C++ does?

**Review by mruberry — APPROVED** [review]
- Should the names be more consistent with the addmm implementation?
- Should we test that the row major inputs receive row major gradients, too?
- Do we need an error for mat1 backward if mat1 is_sparse?

File: torch/csrc/jit/symbolic_script.cpp (hunk 2 above)
Comment (ngimel) [reply torch/csrc/jit/symbolic_script.cpp]: raising RuntimeError fails in interpreter, and this behavior change is beyond the scope of this PR. Sparse support is experimental in script (and a warning to this effect is printed when running with sparse tensors).

**Pull request closed without merging**